pkg/logger performance improvements

aanm opened the pull request:
```release-note
Improved performance of endpoint logging
```

Please check per commit basis

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/5270)
<!-- Reviewable:end -->

aanm committed:
pkg/endpoint replace logger atomically

Signed-off-by: André Martins <andre@cilium.io>

aanm committed:
refactored dockerID to containerID

Signed-off-by: André Martins <andre@cilium.io>

aanm committed:
pkg/endpoint: only update logger fields when changed

Signed-off-by: André Martins <andre@cilium.io>

tgraf commented on pkg/endpoint/log.go:
```diff
@@ -75,16 +78,16 @@ func (e *Endpoint) updateLogger() {
-	// Note: endpoint.loggerMutex protects the reference but not the logger objects. We
-	// cannot update the old object directly as that could be racey.
-	e.logger = baseLogger.WithFields(logrus.Fields{
+	l := baseLogger.WithFields(logrus.Fields{
```
Don't we need to serialize calls to `updateLogger()`? Does it matter?

aanm replied:
I don't think it matters TBH

aanm replied:
@tgraf here

tgraf commented on daemon/endpoint.go:
```diff
@@ -436,6 +436,7 @@ func (h *patchEndpointID) Handle(params PatchEndpointIDParams) middleware.Respon
 		}
 	}
 
+	ep.UpdateLogger(nil)
```
I feel a bit uneasy that we will forget to call this in places. How do developers know when to call this?

aanm replied:
@tgraf 
The `UpdateLogger` is used on all Seters that exist for the endpoint and in there, the UpdateLogger is already being called. Would it help if I add a comment where the logger is being created with the fields?

tgraf replied:
Adding a comment to all fields in the endpoint struct to make sure we don't write to them directly is enough for me. Good to see that this is already abstracted into setters.

rlenglet commented on pkg/endpoint/log.go:
```diff
@@ -29,11 +32,9 @@ var log = logging.DefaultLogger.WithField(logfields.LogSubsys, "endpoint")
 func (e *Endpoint) getLogger() *logrus.Entry {
 	e.updateLogger()
 
-	e.loggerMutex.RLock()
```
Remove the `loggerMutex` field from `Endpoint`. It's not used anywhere anymore.

rlenglet commented on pkg/endpoint/log.go:
```diff
@@ -75,16 +78,16 @@ func (e *Endpoint) updateLogger() {
 	})
-	e.loggerMutex.Unlock()
+
+	// Note: endpoint.loggerMutex protects the reference but not the logger objects. We
```
Remove this comment. `loggerMutex` is not used anymore.

rlenglet commented on pkg/endpoint/endpoint.go:
```diff
@@ -491,19 +491,13 @@ func (e *Endpoint) WaitForProxyCompletions(proxyWaitGroup *completion.WaitGroup)
 
 	start := time.Now()
 
-	if err := e.RLockAlive(); err != nil {
```
This is great!

rlenglet reviewed (approved):
This is great. Just a few nitpicks.

Also, please replace the `logger` local variable in `Endpoint.regenerateBPF` with calls to `e.getLogger()`. The state of the endpoint often changes during regeneration, and using a cached logger makes the logs very confusing when debugging this.

Same for other regeneration methods: `Endpoint.regenerate`, `Endpoint.regeneratePolicy`, `Endpoint.UpdateLabels` and the methods it calls, etc.

tgraf commented on pkg/endpoint/policy.go:
```diff
@@ -693,11 +693,7 @@ func (e *Endpoint) regenerate(owner Owner, reason string) (retErr error) {
 	e.BuildMutex.Lock()
 	defer e.BuildMutex.Unlock()
 
-	if err = e.RLockAlive(); err != nil {
```
Actually, we need to check whether the endpoint is alive after taking the BuildMutex. We don't want to build dead endpoints.

aanm committed:
pkg/endpoint: use e.getLogger() on all log messages

As the state of the endpoint can change during the regenerate, by
calling e.getLogger() every time it will make loggs less confusing
to read.

Signed-off-by: André Martins <andre@cilium.io>
